File: transactions.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const int MON=0,TUE=1,WED=2,THU=3,FRI=4,SAT=5,SUN=6;

int get_weekday(char *s);
void get_nextmonth(char *m, char *nm);

int main() {
	char month[8], weekday_s[8], fstring[255], duemonth[8], nextmonth[8];
	int length, weekday, power, dueday;
	printf("Enter name of month (3 chars): ");
	gets(month);
	month[3] = '\0';
	get_nextmonth(month,nextmonth);
	printf("Enter day of week of 1st of %s (3 chars): ",month);
	gets(weekday_s);
	weekday_s[3] = '\0';
	weekday = get_weekday(weekday_s);
	printf("Enter length of %s in days: ",month);
	scanf("%d",&length);
	system("cls");
	power = 0;
	for(int day=1;day<=length;day++) {
		if(weekday==MON) {
			if((day>7) && (day<=14)) {
				strcpy(fstring,"SlingshotADSL     -59      ~%0.2d %s  ~%0.2d %s  FC\n");
				printf(fstring,day+7,month,day,month);
			}
		}
		if(weekday==TUE) {
			strcpy(duemonth,month);
			dueday = day+1;
			if (dueday > length) {
				dueday-=length;
				strcpy(duemonth,nextmonth);
			}
			strcpy(fstring,"Nathan   Rent     +165      %0.2d %s   %0.2d %s  IN\n");
			printf(fstring,dueday,duemonth,day,month);
			if(day<=7) {
				strcpy(fstring,"Nathan   Power    +??       %0.2d %s   %0.2d %s  IN\n");
				printf(fstring,day+14,month,day,month);
				strcpy(fstring,"Nathan   ADSL     +30       15 %s  ~%0.2d %s  IN\n");
				printf(fstring,month,day,month);
			}
		}
		if(weekday==WED) {
			strcpy(duemonth,month);
			dueday = day+2;
			if (dueday > length) {
				dueday-=length;
				strcpy(duemonth,nextmonth);
			}
			strcpy(fstring,"ChsWlsn  Rent     -360      %0.2d %s   %0.2d %s  AP\n");
			printf(fstring,dueday,duemonth,day,month);
		}
		if(weekday==FRI) {
			if(length-day < 7) {
				// final Friday of month
				power += 35;
				strcpy(duemonth,month);
				dueday = day+2;
				if (dueday > length) {
					dueday-=length;
					strcpy(duemonth,nextmonth);
				}
				strcpy(fstring,"PwrShop  Power    -35       %0.2d %s   %0.2d %s  CC\n");
				printf(fstring,dueday,duemonth,day,month);
				printf("\t(Inc 2 futr pax) (Mth total: %d)\n",power);
			} else {
				// other Fridays
				power += 5;
				strcpy(fstring,"PwrShop  Power    -5        %0.2d %s   %0.2d %s  CC\n");
				printf(fstring,day+2,month,day,month);
			}
		}
		weekday++;
		if(weekday>SUN) weekday=MON;
	}
	printf("==========\n");
}

int get_weekday(char *s) {
	for(unsigned i=0;i<strlen(s);i++) {
		if ((s[i]>='A') && (s[i]<='Z')) {
			s[i] += 'a' - 'A';
		}
	}
	if (strcmp(s,"mon")==0) return MON;
	if (strcmp(s,"tue")==0) return TUE;
	if (strcmp(s,"wed")==0) return WED;
	if (strcmp(s,"thu")==0) return THU;
	if (strcmp(s,"fri")==0) return FRI;
	if (strcmp(s,"sat")==0) return SAT;
	return SUN;
}

void get_nextmonth(char *m, char *nm) {
	if ((m[0]>='a') && (m[0]<='z')) {
		m[0] += 'A' - 'a';
	}
	for(unsigned i=1;i<strlen(m);i++) {
		if ((m[i]>='A') && (m[i]<='Z')) {
			m[i] += 'a' - 'A';
		}
	}
	if (strcmp(m,"Jan")==0) strcpy(nm, "Feb");
	if (strcmp(m,"Feb")==0) strcpy(nm, "Mar");
	if (strcmp(m,"Mar")==0) strcpy(nm, "Apr");
	if (strcmp(m,"Apr")==0) strcpy(nm, "May");
	if (strcmp(m,"May")==0) strcpy(nm, "Jun");
	if (strcmp(m,"Jun")==0) strcpy(nm, "Jul");
	if (strcmp(m,"Jul")==0) strcpy(nm, "Aug");
	if (strcmp(m,"Aug")==0) strcpy(nm, "Sep");
	if (strcmp(m,"Sep")==0) strcpy(nm, "Oct");
	if (strcmp(m,"Oct")==0) strcpy(nm, "Nov");
	if (strcmp(m,"Nov")==0) strcpy(nm, "Dec");
	if (strcmp(m,"Dec")==0) strcpy(nm, "Jan");
}

File: Transactions.java
import java.util.*;


public class Transactions {
	
	public static final String[] MONTH =
			{"Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"};
	public static final int[] MONTHLENGTH = {31,28,31,30,31,30,31,31,30,31,30,31};
	public static final String[] WEEKDAY = {"Mon","Tue","Wed","Thu","Fri","Sat","Sun"};

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		
		String month = "";
		// Input verification loop
		while(!Arrays.asList(MONTH).contains(month)) {
			System.out.print("Enter name of month (3 chars): ");
			month = sc.nextLine().toUpperCase();
			if(month.length()<3) continue; // Avoid index out of bounds exception
			// Set first char to upper case, rest to lower case; truncate to 3 chars
			month = month.substring(0,1) + month.substring(1,3).toLowerCase();
		}
		// Add one to index of this month to get next month; modulus wraps Dec->Jan
		String nextmonth = MONTH[(Arrays.asList(MONTH).indexOf(month)+1) % MONTH.length];
		// System.out.println("month = " + month + ", nextmonth = " + nextmonth);
		
		
		String weekday = "";
		while(!Arrays.asList(WEEKDAY).contains(weekday)) {
			System.out.print("Enter weekday of 1st of " + month + " (3 chars): ");
			weekday = sc.nextLine().toUpperCase();
			if (weekday.length()<3) continue; // Avoid index out of bounds exception
			weekday = weekday.substring(0,1) + weekday.substring(1,3).toLowerCase();
		}		
		// System.out.println("weekday = " + weekday);
		
		int length = MONTHLENGTH[Arrays.asList(MONTH).indexOf(month)];
		if(length == 28) { // February
			String leapyear = "";
			while(!leapyear.equals("Y") && !leapyear.equals("N")) {
				System.out.print("Leap year? (Y/N): ");
				leapyear = sc.nextLine().toUpperCase();
			}
			if(leapyear.equals("Y")) length = 29;
		}
		// System.out.println("length = " + length);
		
		sc.close();
		clearConsole();
		
		int power = 0;
		for(int day=1;day<=length;day++) {
			String fstring, duemonth;
			int dueday;
			int week = ((day-1)/7)+1;
			if(weekday.equals("Mon") && (week==2)) {
				fstring = "SlingshotADSL     -59      ~%02d %s  ~%02d %s  FC%n";
				System.out.format(fstring,day+7,month,day,month);
			}
			if(weekday.equals("Tue")) {
				duemonth = month;
				dueday = day+1;
				if(dueday>length){
					dueday -= length;
					duemonth = nextmonth; 
				}
				fstring = "Nathan   Rent     +185      %02d %s   %02d %s  IN%n";
				System.out.format(fstring,dueday,duemonth,day,month);
				if(week==1){
					fstring = "Nathan   Power    +??       %02d %s   %02d %s  IN%n";
					System.out.format(fstring,day+14,month,day,month);
					fstring = "Nathan   ADSL     +30       15 %s  ~%02d %s  IN%n";
					System.out.format(fstring,month,day,month);
				}
			}
			if(weekday.equals("Wed")) {
				duemonth = month;
				dueday = day+2;
				if(dueday>length) {
					dueday -= length;
					duemonth = nextmonth;
				}
				fstring = "ChsWlsn  Rent     -400      %02d %s   %02d %s  AP%n";
				System.out.format(fstring,dueday,duemonth,day,month);
			}
			if(weekday.equals("Fri")) {
				if(length-day < 7) {
					// final Friday of month
					power += 35;
					duemonth = month;
					dueday = day+2;
					if(dueday>length) {
						dueday -= length;
						duemonth = nextmonth;
					}
					fstring = "PwrShop  Power    -35       %02d %s   %02d %s  CC%n";
					System.out.format(fstring,dueday,duemonth,day,month);
					System.out.format("        (Inc 2 futr pax) (Mth total: %d)%n",power);
				} else {
					// other Fridays
					power += 5;
					fstring = "PwrShop  Power    -5        %02d %s   %02d %s  CC%n";
					System.out.format(fstring,day+2,month,day,month);
				}
			}
			// increment weekday
			weekday = WEEKDAY[(Arrays.asList(WEEKDAY).indexOf(weekday)+1) % WEEKDAY.length];
		}
	}
	
	public final static void clearConsole() {
	    try {
	        final String os = System.getProperty("os.name");
	        if (os.contains("Windows")) {
	            Runtime.getRuntime().exec("cls");
	        } else {
	            Runtime.getRuntime().exec("clear");
	        }
	    }
	    catch (final Exception e) {}
	}
	
}
